### ISSUE TITLE
Support for face and edge quadrature

### ISSUE THREAD

**Issue by thomasmelvin** [stfc/PSyclone#193]
LFRic [#1506](https://code.metoffice.gov.uk/trac/lfric/ticket/1506) Introduced two new quadrature objects  (which are also used by LFRic [#1449](https://code.metoffice.gov.uk/trac/lfric/ticket/1449) ):

- `quadrature_face_type`
- `quadrature_edge_type`

For computing quadrature on the faces and edges of cells respectively (instead of just in the volume).
These are built in a similar way the existing `quadrature_xzy_type`. 

The existing kernel metadata has been extended to allow:
- `integer :: gh_shape = GH_QUADRATURE_face`
- `integer :: gh_shape = GH_QUADRATURE_edge`

In which case a quadrature of the appropriate type will be passed in though the invoke.

The psy-layer code that needs to be generated is:
```fortran
  type(quadrature_face_type), intent( in ) :: qr
  type(quadrature_face_proxy_type) :: qr_proxy
  real(kind=r_def), allocatable :: face_basis(:,:,:,:)
  integer :: nqp_f, nfaces
  real(kind=r_def), pointer ::  wqp_f(:,:) => null()

  qr_proxy = qr%get_quadrature_proxy()
  nfaces = qr_proxy%nfaces
  nqp_f  = qr_proxy%np_xyz
  wqp_f => qr_proxy%weights_xyz

  ! Evaluate the basis functions on the faces
  allocate( face_basis(dim_wx, ndf_wx, nqp_f, nfaces))
  call qr%compute_function(BASIS, chi_proxy(1)%vspace, dim_wx, ndf_wx, face_basis)
```
For example see `invoke_ru_bd_kernel` in `psykal_lite_mod.F90`

The same code is required for `quadrature_edge_type`  with '_face' replaced by '_edge'

**Comment by arporter** [comment]
I see from looking at `invoke_ru_bd_kernel` that we will also require #18 to be done! Do we need to support differential basis functions for these new quadrature types too. If so, do the associated arrays have the same dimensions?

**Comment by thomasmelvin** [comment]
Yes we will also need differential basis and you're right they will have the same size (apart from `dim` being replaced by `diff_dim`)

**Comment by arporter** [comment]
Thanks @thomasmelvin. I'm just looking at the ru_bd_kernel and see that the last dimension of the basis arrays and the weights array is hardwired to 4:

    real(kind=r_def), dimension(3,ndf_w2,nqp,4),     intent(in) :: w2_basis_face
    real(kind=r_def), dimension(1,ndf_w3,nqp,4),     intent(in) :: w3_basis_face
    real(kind=r_def), dimension(1,ndf_wtheta,nqp,4), intent(in) :: wtheta_basis_face
    ...
    real(kind=r_def), dimension(nqp,4), intent(in)      ::  wqp
 
Is that the way you want it or should we really pass nfaces into the kernel and use that instead?

**Comment by thomasmelvin** [comment]
@arporter I guess we should really pass `nfaces` in, I had treated it like `dimension` which we don't pass in. But if we wanted to generically support triangles (or other shapes) then we should pass it in

**Comment by arporter** [comment]
OK, I shall pass it in.

**Comment by arporter** [comment]
I have a lot of the basics working now (it turns out that the existing QR support is easy to extend). However, we need the reference element in order to get the number of edges/faces and we have #18 that needs to implement that too. Therefore I'm going to pause this issue and switch to #18.

**Comment by arporter** [comment]
Actually, #18 seems to have been superseded by #150 so that's the one I'll do. (@thomasmelvin - does this mean I can close #18?)

**Comment by thomasmelvin** [comment]
Yes, #18 can be closed as duplicate of #150

**Comment by arporter** [comment]
#150 is nearly done so I'm reviving this Issue. I've brought the branch up-to-date with master.

**Comment by arporter** [comment]
The PSyKAlite code on LFRic master gets the number of faces/edges from the quadrature object. I know @TeranIvy has talked about this before but the question is (for @TeranIvy or @thomasmelvin), should PSyclone get these quantities from the quadrature object or from the reference element?

**Comment by TeranIvy** [comment]
PSyclone should get the number of faces for face/edge quadrature from the quadrature object, see [psykal_lite_mod.F90](https://code.metoffice.gov.uk/trac/lfric/browser/LFRic/branches/dev/ivakavcic/r21216_RefElemArrays/gungho/source/psy/psykal_lite_mod.F90) in my branch [r21216_RefElemArrays](https://code.metoffice.gov.uk/trac/lfric/browser/LFRic/branches/dev/ivakavcic/r21216_RefElemArrays) ([LFRic ticket #1916](https://code.metoffice.gov.uk/trac/lfric/ticket/1916)). The reference element arrays should get the number of faces from the reference element object (also in PSyKAl-lite). 

Note that the face quadrature getters do not posses knowledge about whether the number of faces they get is horizontal or vertical as that is decided during the face quadrature constructor phase (see [quadrature_face_mod.F90](https://code.metoffice.gov.uk/trac/lfric/browser/LFRic/branches/dev/ivakavcic/r21216_RefElemArrays/infrastructure/source/field/quadrature_face_mod.F90) in the branch). This is why we agreed to pass both the number of faces from quadrature and the number of horizontal/vertical faces from the reference element if both are present. In practice it is very unlikely that horizontal face quadrature will not be paired with the horizontal reference element arrays, however this is not visible from getters (comments in the PSyKAl-lite code in my branch). The downside is that we will probably pass duplicate information for now. The upside is that the support for the face quadrature can be developed independently from the support for the reference element.

**Comment by arporter** [comment]
Thanks @TeranIvy for the heads-up about the change on LFRic master. I've had a look at `./gungho/source/kernel/rtheta_bd_kernel_mod.F90` and I need to make a couple of changes to what I have. However, I also note that that kernel has the basis-function arrays separated from the other function-space-related arguments which is not consistent with the current API rules. The kernel also talks about 'stencil' information but this is not mentioned in the meta-data (unless I'm missing something)?

**Comment by thomasmelvin** [comment]
It looks to me that both the meta-data (missing stencil information) and the argument list (inconsistent order with the current api) for that kernel are incorrect, these will both need updating before we can remove the pyskalite code for it

**Comment by TeranIvy** [comment]
As @thomasmelvin says, we will need to update metadata and reorder the arguments - it is now possible to have the incorrect metadata because the `_bd_` kernels are not "psycloned". The [psy-kal](https://code.metoffice.gov.uk/trac/lfric/browser/LFRic/trunk/gungho/source/psy/psykal_lite_mod.F90) lite code should hopefully shed more light.
There is also an old example of `rtheta_bd_kernel_mod.F90` in PSyclone `examples/lfric/eg9`. I added it as a part of another issue and changed `face` to `xyoz` quadrature for examples to work. The metadata for the reference element, quadrature and mesh need to be updated as well. Perhaps this would be a good test case that the quadrature support works?

**Comment by TeranIvy** [comment]
Closed via #701
